### ISSUE TITLE
[Feature] Making UnityAtoms compatible with Enter The PlayMode feature of Unity

### ISSUE THREAD

**Issue by mnarimani** [AdamRamberg/unity-atoms#146]
Enter The PlayMode is a feature that makes entering the play mode insanely fast. 
![image](https://user-images.githubusercontent.com/30625116/80925172-eede8400-8da2-11ea-8e1c-4002a6bed917.png)

But there are side effects:

- Fields marked with [NonSerialized] attribute keep their values - as script serialization step is skipped completely and scripts remain the same, all data is carried to the Play Mode.
- Static variables keep their values - because Domain Reload does not happen types stay initialized.
- Static events keep their subscribers.
- There is no extra OnDisable/OnEnable calls for ExecuteInEditMode scripts.
- ScriptableObject, MonoBehaviour fields which are not serialized into the build ([NonSerialized], private or internal) keep their values - as existing objects are not recreated and constructors are not called.
- Watch out for null private and internal fields of array/List type - they are converted to an empty array/List object during Domain Reload and stay non null for game scripts.
- ExecuteInEditMode/ExecuteAlways scripts are not destroyed/awaken - no OnDestroy/Awake calls for those.
- Watch out for Awake/OnEnable methods which check EditorApplication.isPlaying property - Awake is not called and OnEnable is called only when EditorApplication.isPlaying is already true on Play Mode change.

UnityAtoms gets heavily impacted by these side effects.
**OnEnable and OnDisable calls are not called anymore.** So nothing works as expected.

I exactly don't know how we are supposed to change the code at the moment. So I created this issue for the discussion about it.

There are some useful callbacks that we can use. 

- `EditorApplication.playModeStateChanged` event
- ` [RuntimeInitializeOnLoadMethod]` attribute

What do you think?

**Comment by mnarimani** [comment]
I think `bug` label needs to be changed into something more appropriate. I mistakenly created this out of bug report template.

**Comment by AdamRamberg** [comment]
Thanks for creating this issue! 

I'm completely new to this Unity feature, but on the surface this seems to be a huge undertaking, but I could be wrong. Do you know where to start to make this happen? Are there for example any migration guides out there?  

I'm all for shaving off time when entering play mode. If you (or anyone else) wants to contribute to Atoms and give this a try then I would be very grateful. Not sure that I will have the time to implement this anytime soon though.

**Comment by mnarimani** [comment]
The main issue is `OnEnable` and `OnDisable` calls. I have not found a way to get the same effect with Enter the Playmode yet. Every guide that is out there is for handling static events and variables. (ScriptableObjects always get neglected in Unity. lol)
When I find a way, I will update this issue.

**Comment by soraphis** [comment]
Maybe falling back to https://docs.unity3d.com/ScriptReference/ISerializationCallbackReceiver.OnBeforeSerialize.html ?

i know this had different problems, but they might be easier to solve?

**Comment by hazarartuner** [comment]
Hi all, I've created a pull request to workaround this issue: https://github.com/AdamRamberg/unity-atoms/pull/182 

We do not have lots of options therefore it may not seem optimal solution but it seems work. I'd like to discuss about the solution and find better approach if possible.

**Comment by TowerCrow** [comment]
Start Cleared option on ValueList also doesn't work with the new Enter Play Mode settings.

**Comment by hazarartuner** [comment]
> Start Cleared option on ValueList also doesn't work with the new Enter Play Mode settings.

I didn't use `Start Cleared` option before and I don't know about this feature. I'll look at this issue in a new PR when I'm available and I guess it will be next week :)

**Comment by tarragonfly** [comment]
I'm not sure what the current progress on supporting this is, but I have issues with Event Reference Listeners triggering as soon as I enter play mode with the new Enter Play Mode settings - both domain and scene reload disabled. 

I can manually fix this by either setting Replay Buffer Size to 0 or by disabling the Replay Event Buffer on Register bool.

**Comment by rhys-vdw** [comment]
There should be a warning at the start of the docs about this. I just spent two days implementing this library before starting to hit errors when re-entering play mode.

**Comment by soraphis** [comment]
thanks for bringing this to my attention.
could you give a bit more details what errors you run into, and maybe a simple setup to reproduce them?

**Comment by rhys-vdw** [comment]
I am confused as to why 'Old Value' is serialized though... Is this really information that needs to be saved to the project? I feel like "old value" and "value" should only exist at runtime, (as opposed to "Initial Value" which obviously needs to be saved).

![image](https://user-images.githubusercontent.com/822105/215663240-d2019280-b110-4d0c-bcef-6c515fdcb783.png)

Not sure if this is related to domain reload or not.

**Comment by AdamRamberg** [comment]
Been trying to come back to this issue and wrap my head around it. This might be obvious, but wasn't clear to me when reading Unity's docs on what it means to disable domain reloading and how and why @hazarartuner solution works. Please correct me if any of the below statements are wrong. 

- `OnDisable()` and `OnEnable()` are not called anymore when entering play mode since it is part of domain reloading. 
- `Awake()` and `OnEnable()` are still being called for SOs when starting up the editor. 
- @hazarartuner works, but only if we restart the editor, since `OnEnable()` is called for all SOs in the project when starting the editor. 

Could anyone point me to where in the Unity docs it is stated that `OnDisable()` and `OnEnable()` is part of domain reloading and is therefore not called when entering play mode? There are only notes for it when used in scripts using `ExecuteInEditMode`. Had to read [this](https://forum.unity.com/threads/scriptableobject-awake-onenable-ondisable-ondestroy-demystified.1415565/) thread to grasp this, but there must be some officilar resource on this?

**Comment by rhys-vdw** [comment]
@AdamRamberg 

> Could anyone point me to where in the Unity docs it is stated that OnDisable() and OnEnable() is part of domain reloading and is therefore not called when entering play mode?

You can see in the SO docs:

[`OnEnable`](https://docs.unity3d.com/ScriptReference/ScriptableObject.OnEnable.html):
> This function is called when the object is loaded.

This is not explicit, but assets are (as I understand) loaded on demand by the editor, either when selected for inspector or used in game. They are not reloaded when entering play mode (hence why modifications to the asset persist when entering play mode).

[`OnDisable`](https://docs.unity3d.com/ScriptReference/ScriptableObject.OnDisable.html):
> This function is called when the scriptable object goes out of scope.
>
> This is also called when the object is destroyed and can be used for any cleanup code. When scripts are reloaded after compilation has finished, OnDisable will be called, followed by an OnEnable after the script has been loaded.

And here it is clearly. OnDisable is called when cleaning up SOs, and OnEnable is called after loading.

If you want to trigger code when entering playmode, I'd recommend looking at [`RuntimeInitializeOnLoadMethodAttribute`](https://docs.unity3d.com/ScriptReference/RuntimeInitializeOnLoadMethodAttribute.html). I used this to clear out pools and reset static fields in our project. However we have a single scene project, so I'm not sure how it plays in multi-scene.

However, a cleaner fix may simply be to not serialize `_value` and `_oldValue` at all as I hinted at [above](https://github.com/unity-atoms/unity-atoms/issues/146#issuecomment-1409735010).

**Comment by rhys-vdw** [comment]
Just to go into more detail on the `[RuntimeInitializeOnLoadMethodAttribute](https://docs.unity3d.com/ScriptReference/RuntimeInitializeOnLoadMethodAttribute.html)` idea:

I think you could have a single static handler that calls [`Resources.FindObjectsOfTypeAll`https://docs.unity3d.com/ScriptReference/Resources.FindObjectsOfTypeAll.html] like:

```cs
[RuntimeInitializeOnLoadMethodAttribute]
class ResetAll() {
  // get all atoms currently in memory.
  var atoms = Resources.FindObjectsOfTypeAll(typeof(IAtomBaseInterface)) as IAtomBaseInterface[];
  // reset current value to initial value.
  foreach (var a in atoms) a.Reset();
}
```

Might be more appropriate to use `EditorApplication.playModeStateChanged` (as @pnarimani suggested) since exiting playmode is also important.

However... I still feel these values just shouldn't be serialized at all unless I'm missing something.

**Comment by soraphis** [comment]
> Please correct me if any of the below statements are wrong.

> [ScriptableObject.]OnDisable() and [ScriptableObject.]OnEnable() are not called anymore when entering play mode since it is part of domain reloading.

obviously, but for completeness sake: for runtime instantaited SOs, they should still be called.
also, I'm not entirely sure about AssetBundles/Addressables in edit mode, as those would have their own lifecycle that differs a bit from the usual editor-assets-lifecycle


> @hazarartuner works, but only if we restart the editor, since OnEnable() is called for all SOs in the project when starting the editor.

also compiling any code change should trigger a domain reload I think. 

> Could anyone point me to where in the Unity docs it is stated that OnDisable() and OnEnable() is part of domain reloading and is therefore not called when entering play mode?

It's not _really_ part of domain reloading, but a consequence of it. ScriptableObject.OnEnable states "This function is called when the object is loaded.", without a domain reload, the asset just stays to be loaded, as there is nothing that unloads them. 

But the whole ScriptableObject/Asset lifecycle is kinda a neglected topic in the docs, hence it often results in confusion of atoms-users when the things behave differently in a build compared to the editor.

___

I also agree with everything @rhys-vdw stated.

**Comment by AdamRamberg** [comment]
Thanks a bunch both of you for the explanations 🙏

I agree, makes sense what you @rhys-vdw are suggestiong.

**Comment by soraphis** [comment]
I'd say the way atoms are resetted in general should be changed to use fully utilize the RuntimeInitializeOnLoad approach 

https://github.com/unity-atoms/unity-atoms/issues/146#issuecomment-1641273531

together with things like https://github.com/unity-atoms/unity-atoms/issues/387 this should be a bit cleaner on the kept-memory between sessions and cleaner than relying on OnEnable an things like that. 

(this comment is more for my future self to summarize my thoughts right now)

**Comment by soraphis** [comment]
I'm closing this right now. Unity Atoms IS compatible with enabled and disabled domain reload.
We're not using RuntimeInitializeOnLoad so far, progress on this topic will be tracked https://github.com/unity-atoms/unity-atoms/issues/496
